{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A notebook to develop the script, generate and analyse some results.\n",
    "\n",
    "I have started the development of the program in this notebook. Then, I have adapted the code to have a script\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "'''\n",
    "Created on \n",
    "\n",
    "@author: manuel.dominguezbecerra@nhs.net\n",
    "\n",
    "'''\n",
    "\n",
    "# Import libraries \n",
    "import argparse                          # pip install argparse\n",
    "import pandas as pd                      \n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "# To allow the user to load the file after running \"python run.ppy\"\n",
    "\n",
    "# Example:    python run.py -- /path/file_name1 -- /path/file_name2\n",
    "\n",
    "   ### 1. Load the files\n",
    "    \n",
    "# During the development of this application, I have needed to run the script many times\n",
    "# To do this in an easy way, I directly introduce here the path/file_name in a variable\n",
    "# As it can be seen below\n",
    "\n",
    "Name_file1 = \"/Users/monkiky/Desktop/VCF-matcher/Twin_brothers/W1717693_S1.vcf\"\n",
    "\n",
    "Name_file2 = \"/Users/monkiky/Desktop/VCF-matcher/Twin_brothers/W1816869_S14.vcf\"\n",
    "\n",
    "# In the script, the previous two lines doesn´t appear and the following lines are incommeneted\n",
    "\n",
    "#parser = argparse.ArgumentParser()\n",
    "#parser.add_argument('--vcf1', type=str, required=True)\n",
    "#parser.add_argument('--vcf2', type=int, required=True)\n",
    "#args = parser.parse_args()\n",
    "#Name_file1 = args.vcf1 \n",
    "#Name_file2 = args.vcf2\n",
    "\n",
    "# This 6 lines allow the user to select the two neccesary files to run the script\n",
    "\n",
    "#During the development, the few lines make running the script easier.\n",
    "# In the run.py script, the previous lines are uncommented and the following lines are deleted.\n",
    "\n",
    "\n",
    "# From here to the end, the following lines are identical that the one find in the script\n",
    "\n",
    "\n",
    "   ### 2. Take the body of the files and ignore meta-data lines\n",
    "\n",
    "# Originally what I did was to delete the metedata and save the data in a new file\n",
    "# This approach was over complicated and doesn´t work in window machines.\n",
    "# Instead of this, the current approach  take directly the data and save it in pandas dataframe object.\n",
    "# This is also a more efficient method.\n",
    "\n",
    "# For the file number 1\n",
    "with open(Name_file1, 'r') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('#') and len(line)>2 and line[1] != '#':\n",
    "            columns = line[1:-1].split('\\t')\n",
    "            break\n",
    "\n",
    "dataA = pd.read_csv(Name_file1, comment='#', delimiter='\\t', names=columns)\n",
    "\n",
    "# Same for the second file\n",
    "with open(Name_file2, 'r') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('#') and len(line)>2 and line[1] != '#':\n",
    "            columns = line[1:-1].split('\\t')\n",
    "            break\n",
    "            \n",
    "dataB = pd.read_csv(Name_file2, comment='#', delimiter='\\t', names=columns)\n",
    "\n",
    "\n",
    "# If the data frame has more than 10 columns\n",
    "# that means there are more than 1 sample in the vcf file\n",
    "# If more than one file is found, next lines ask to the user what sample to take.\n",
    "\n",
    "if len(dataA.columns) > 10:\n",
    "     Name_sample1 = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(Name_file1)))\n",
    "else:\n",
    "    print('File', os.path.basename(Name_file1), 'contains one sample only.')\n",
    "    Name_sample1 = dataA.columns.tolist()[9]\n",
    "\n",
    "if len(dataB.columns) > 10:\n",
    "     Name_sample2 = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(Name_file2)))\n",
    "else:\n",
    "    print('File', os.path.basename(Name_file2), 'contains one sample only.')\n",
    "    Name_sample2 = dataB.columns.tolist()[9]\n",
    "\n",
    "    \n",
    "\n",
    "# The data of the sample is save in the file in one column separated by \":\". \n",
    "# Example: Example:   0/1:100:3501,254:0.0676:20:-100.0000:100\n",
    "\n",
    "# To take the data I want, I need to split and put the fiels in columns\n",
    "\n",
    "\n",
    "tmp_df2 = dataA[Name_sample1].str.split(':', expand=True)\n",
    "\n",
    "#tmp_df2 = tmp_df2.fillna(value=np.nan)\n",
    "#tmp_df2 = tmp_df2.dropna(axis=1, how='any')\n",
    "\n",
    "tmp_df2.columns = dataA.FORMAT.iloc[0].split(':')\n",
    "dataA = pd.concat([dataA, tmp_df2], axis=1)\n",
    "\n",
    "# Results so far\n",
    "#  CHROM       POS ID REF ALT  QUAL FILTER   GT  GQ       AD      VF  NL\n",
    "#  chr1  36931745  .   G   A  34.0     SB  0/1  34  1755,34  0.0190  20\n",
    "\n",
    "# Same for dataB\n",
    "tmp_df3 = dataB[Name_sample2].str.split(':', expand=True)\n",
    "tmp_df3 = tmp_df3.fillna(value=np.nan) \n",
    "tmp_df3 = tmp_df3.dropna(axis=1, how='any') \n",
    "\n",
    "tmp_df3.columns = dataB.FORMAT.iloc[0].split(':')\n",
    "dataB = pd.concat([dataB, tmp_df3], axis=1)\n",
    "\n",
    "\n",
    "\n",
    "    ### 3. Filter the variants  \n",
    "    \n",
    "# Two filters are applied here. First, I take only the variants that \"PASS\" the \"FILTER\" field. \n",
    "# This is a mandatory field.\n",
    "\n",
    "# Then, I take only the variants which Variant Frecuency (VF) is highter than. 0.4\n",
    "# This is not a mandatory field in vcf filed\n",
    "\n",
    "# VF is in the 9th column with another data. \n",
    "#we need to extract and put VF data in an independent column first\n",
    "\n",
    "\n",
    "# Now, let´s filter the variant\n",
    "# We dont want variants that doesnt pass the Filter status\n",
    "# This field is mandatory, no condition is needed because it is applied to any type of VCF file.\n",
    "dataA = dataA.query('FILTER == \"PASS\"')\n",
    "dataB = dataB.query('FILTER == \"PASS\"')\n",
    "                    \n",
    "# Now, I apply Variant Frecuency (VF) filter (non-mandatory field)\n",
    "\n",
    "# If there is a VF Genotype fields in the file\n",
    "# Take the variants that are > 0.4\n",
    "# If not, no VF filter is applied\n",
    "\n",
    "if \"VF\" in dataA.columns:\n",
    "    dataA[\"VF\"] = dataA[\"VF\"].astype(float)\n",
    "    dataA = dataA.query('VF > 0.4')\n",
    "    \n",
    "if \"VF\" in dataB.columns:\n",
    "    dataB[\"VF\"] = dataB[\"VF\"].astype(float)\n",
    "    dataB = dataB.query('VF > 0.4')\n",
    "    \n",
    "\n",
    "   ### 4. Variants comparison\n",
    "\n",
    "# Two variants are the same if  CHROM, POS, REF, ALT and GT are equal\n",
    "\n",
    "dataA['CHROMPOSREFALT']= dataA[\"CHROM\"].apply(str)+\".\"+dataA[\"POS\"].apply(str)+dataA[\"REF\"]+\".\"+dataA[\"ALT\"]\n",
    "\n",
    "dataB['CHROMPOSREFALT']= dataB[\"CHROM\"].apply(str)+\".\"+dataB[\"POS\"].apply(str)+dataB[\"REF\"]+\".\"+dataB[\"ALT\"]\n",
    "\n",
    "\n",
    "frames = [dataA[['CHROMPOSREFALT', 'GT']],dataB[['CHROMPOSREFALT', 'GT']]]\n",
    "semi_final_df = pd.concat(frames)\n",
    "# semi_final_df saves the CHROM, POS, REF, ALT in the first column and the GT in the second column\n",
    "\n",
    "\n",
    "#Example:\n",
    "'''\n",
    "CHROMPOSREFALT     GT\n",
    "\n",
    "chr1.123456T.C     0/1       From sample 1\n",
    "chr2.123456A.C     1/1       From sample 1\n",
    "chr1.123456T.C     0/1       From sample 2\n",
    "\n",
    "'''\n",
    "\n",
    "\n",
    "# Let´s reduce the number of rows\n",
    "final_df=(semi_final_df.assign(key=semi_final_df.groupby('CHROMPOSREFALT').cumcount())\n",
    "      .pivot('CHROMPOSREFALT','key','GT')\n",
    "      .rename(columns=lambda x:f\"Sample{x+1}\")\n",
    "      .rename_axis(columns=None).reset_index())\n",
    "\n",
    "# Final_df contains all common and non-common variants that passed the filters\n",
    "\n",
    "# The pivot command gives NaNs in the column of the second sample when a variant is only found in one of the samples\n",
    "# This NaN values return errors in the following lines. This is an error I have found when the script was developed\n",
    "# To solve the issue without introducing many changes, NaN values are replace for 99/99 values.\n",
    "final_df = final_df.replace(np.NaN,\"99/99\" )\n",
    "\n",
    "   ### 4. See how many GT match\n",
    "\n",
    "# Variables needed for both type of reports\n",
    "\n",
    "r0 = os.path.basename(Name_file1) # Get the  file name of a path/file_name \n",
    "r1 = Name_sample1                 # Get the name of the sample 1\n",
    "r2 = os.path.basename(Name_file2)\n",
    "r3 = Name_sample2\n",
    "\n",
    "if \"Sample2\" not in final_df.columns:\n",
    "    # If Sample2 is not in final_df, that means there is not common position between the samples.\n",
    "    report0 = '''\n",
    " _____________________________  REPORT  ________________________________________ \n",
    "\n",
    "vcf 1: {0}  AND its sample name: {1}  \n",
    "vcf 2:  {2}  AND its sample name: {3}\n",
    "\n",
    "   No common positions found between samples\n",
    "\n",
    " ____________________________ END REPORT  _______________________________________\n",
    "'''\n",
    "    print(report0.format(r0,r1,r2,r3,))\n",
    "\n",
    "else:    \n",
    "    # else match common position and carry on the report.\n",
    "    final_df[\"Matches\"] = np.where(final_df[\"Sample1\"] == final_df[\"Sample2\"], True, False)\n",
    "    final_df.columns = ['CHROMPOSREFALT',os.path.basename(Name_file1),os.path.basename(Name_file2),\"Matches\" ]\n",
    "\n",
    "    # 6.  To get hom and het \n",
    "\n",
    "    final_df['Genotype1'] = final_df.iloc[:,2].apply(lambda x: x.split('/' or '|')[0])\n",
    "    final_df['Genotype2'] = final_df.iloc[:,2].apply(lambda x: x.split('/' or '|')[1])\n",
    "    \n",
    "    final_df['Hom/het'] = np.select([final_df['Matches'] & final_df['Genotype1'].eq(final_df['Genotype2']),\n",
    "                       final_df['Matches'] & final_df['Genotype1'].ne(final_df['Genotype2'])],\n",
    "                       choicelist=[\"Hom\", \"Het\"],\n",
    "                       default=pd.NA)\n",
    "\n",
    "   ### 6. Generate results in the terminal\n",
    "\n",
    "# Variable to introduce in the report :\n",
    "    \n",
    "    r4 = final_df['Matches'].value_counts().get(True, 0) # Count trues if any, retunr 0\n",
    "    r5 = final_df['Hom/het'].value_counts().get(\"Hom\", 0)\n",
    "    r6 =final_df['Hom/het'].value_counts().get(\"Het\", 0)\n",
    "    r7 = final_df['Matches'].value_counts().get(False, 0)\n",
    "    r8 = len(final_df)\n",
    "    r9 = r4/(r4+r7)\n",
    "    report = '''\n",
    " _____________________________  REPORT  ________________________________________ \n",
    "\n",
    "vcf 1: {0}  AND its sample name: {1}  \n",
    "vcf 2:  {2}  AND its sample name: {3}\n",
    "\n",
    "                                                  Homozigous: {5} \n",
    "Number of positions with the same genotype: {4} \n",
    "                                                  Heterozigous: {6} \n",
    "                                                \n",
    "                                                  \n",
    "Number of positions with different genotype: {7} \n",
    "                                                  \n",
    "\n",
    "\n",
    "Total positions compared: {8}\n",
    "Percentage in common: {4}/{8}= {9}\n",
    "\n",
    " ____________________________ END REPORT  _______________________________________\n",
    "'''\n",
    "    print(report.format(r0,r1,r2,r3,r4,r5,r6,r7,r8,r9))\n",
    "    \n",
    "    # ========== Results ================== #\n",
    "\n",
    "# To store results in lists and then in df\n",
    "\n",
    "\n",
    " ### This is to create the scatter plot with the matched and unmatched samples\n",
    "\n",
    "# To create the list the first iteration\n",
    "#sampleA = [\"hola\"]\n",
    "#sampleB = [\"hola\"]\n",
    "#Total_positions_comp = [\"hola\"]\n",
    "#Pos_same_genotype = [\"hola\"]\n",
    "#Pos_different_genotyope = [\"hola\"]\n",
    "#Hom = [\"hola\"]\n",
    "#Het = [\"hola\"]\n",
    "\n",
    "#sampleA.append(r1)\n",
    "#sampleB.append(r3)\n",
    "#Total_positions_comp.append(r8)\n",
    "#Pos_same_genotype.append(r4)\n",
    "#Pos_different_genotyope.append(r7)\n",
    "#Hom.append(r5)\n",
    "#Het.append(r6)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_sample (Name_file):\n",
    "    with open(Name_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#') and len(line)>2 and line[1] != '#':\n",
    "                columns = line[1:-1].split('\\t')\n",
    "                break\n",
    "    return columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse                          # pip install argparse\n",
    "import pandas as pd                      \n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "NAME_FILE_1 = \"/Users/monkiky/Desktop/VCF-matcher/Twin_brothers/W1717693_S1.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " 'FORMAT',\n",
       " 'W1717693']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_sample (NAME_FILE_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>W1717693</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36931706</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>32.0</td>\n",
       "      <td>SB</td>\n",
       "      <td>DP=1055;TI=NM_172313,NM_156039,NM_000760;GI=CS...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:32:1031,23:0.0218:20:0.0000:32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36932103</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>62.0</td>\n",
       "      <td>SB</td>\n",
       "      <td>DP=2094;TI=NM_172313,NM_156039,NM_000760;GI=CS...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:62:2031,47:0.0224:20:0.0000:62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36932108</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>43.0</td>\n",
       "      <td>SB</td>\n",
       "      <td>DP=2136;TI=NM_172313,NM_156039,NM_000760;GI=CS...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:43:2083,42:0.0197:20:0.0000:43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36932121</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>65.0</td>\n",
       "      <td>SB</td>\n",
       "      <td>DP=321;TI=NM_172313,NM_156039,NM_000760;GI=CSF...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:65:305,16:0.0498:20:-0.2685:65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>36932122</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>100.0</td>\n",
       "      <td>SB</td>\n",
       "      <td>DP=333;TI=NM_172313,NM_156039,NM_000760;GI=CSF...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:100:304,26:0.0781:20:-0.2588:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133527688</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>100.0</td>\n",
       "      <td>R8</td>\n",
       "      <td>DP=1090;TI=NM_032335,NM_001015877,NM_032458;GI...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:100:1000,90:0.0826:20:-100.0000:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133527928</td>\n",
       "      <td>.</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>50.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=1395;TI=NM_032335,NM_001015877,NM_032458;GI...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:50:1362,33:0.0237:20:-22.4949:50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133547693</td>\n",
       "      <td>.</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=4576;TI=NM_032335,NM_001015877,NM_032458;GI...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:100:4396,180:0.0393:20:-100.0000:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133547693</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=4493;TI=NM_032335,NM_001015877,NM_032458;GI...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:100:4368,125:0.0278:20:-100.0000:100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>chrX</td>\n",
       "      <td>133547816</td>\n",
       "      <td>.</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>100.0</td>\n",
       "      <td>PASS</td>\n",
       "      <td>DP=979;TI=NM_032335,NM_001015877,NM_032458;GI=...</td>\n",
       "      <td>GT:GQ:AD:VF:NL:SB:GQX</td>\n",
       "      <td>0/1:100:943,36:0.0368:20:-48.6825:100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>459 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM        POS ID REF ALT   QUAL FILTER  \\\n",
       "0    chr1   36931706  .   T   G   32.0     SB   \n",
       "1    chr1   36932103  .   T   G   62.0     SB   \n",
       "2    chr1   36932108  .   A   C   43.0     SB   \n",
       "3    chr1   36932121  .   C   G   65.0     SB   \n",
       "4    chr1   36932122  .   T   G  100.0     SB   \n",
       "..    ...        ... ..  ..  ..    ...    ...   \n",
       "454  chrX  133527688  .   G  GT  100.0     R8   \n",
       "455  chrX  133527928  .  AT   A   50.0   PASS   \n",
       "456  chrX  133547693  .  TA   T  100.0   PASS   \n",
       "457  chrX  133547693  .   T  TA  100.0   PASS   \n",
       "458  chrX  133547816  .  AT   A  100.0   PASS   \n",
       "\n",
       "                                                  INFO                 FORMAT  \\\n",
       "0    DP=1055;TI=NM_172313,NM_156039,NM_000760;GI=CS...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "1    DP=2094;TI=NM_172313,NM_156039,NM_000760;GI=CS...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "2    DP=2136;TI=NM_172313,NM_156039,NM_000760;GI=CS...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "3    DP=321;TI=NM_172313,NM_156039,NM_000760;GI=CSF...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "4    DP=333;TI=NM_172313,NM_156039,NM_000760;GI=CSF...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "..                                                 ...                    ...   \n",
       "454  DP=1090;TI=NM_032335,NM_001015877,NM_032458;GI...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "455  DP=1395;TI=NM_032335,NM_001015877,NM_032458;GI...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "456  DP=4576;TI=NM_032335,NM_001015877,NM_032458;GI...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "457  DP=4493;TI=NM_032335,NM_001015877,NM_032458;GI...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "458  DP=979;TI=NM_032335,NM_001015877,NM_032458;GI=...  GT:GQ:AD:VF:NL:SB:GQX   \n",
       "\n",
       "                                     W1717693  \n",
       "0          0/1:32:1031,23:0.0218:20:0.0000:32  \n",
       "1          0/1:62:2031,47:0.0224:20:0.0000:62  \n",
       "2          0/1:43:2083,42:0.0197:20:0.0000:43  \n",
       "3          0/1:65:305,16:0.0498:20:-0.2685:65  \n",
       "4        0/1:100:304,26:0.0781:20:-0.2588:100  \n",
       "..                                        ...  \n",
       "454   0/1:100:1000,90:0.0826:20:-100.0000:100  \n",
       "455      0/1:50:1362,33:0.0237:20:-22.4949:50  \n",
       "456  0/1:100:4396,180:0.0393:20:-100.0000:100  \n",
       "457  0/1:100:4368,125:0.0278:20:-100.0000:100  \n",
       "458     0/1:100:943,36:0.0368:20:-48.6825:100  \n",
       "\n",
       "[459 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataA = pd.read_csv(NAME_FILE_1, comment='#', delimiter='\\t', names=load_sample (NAME_FILE_1))\n",
    "dataA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def take_sample(data,NAME_FILE):\n",
    "    '''\n",
    "    VCF files can have 1 or > 1 samples\n",
    "    We need to know what sample the user wants to compare\n",
    "    After this, the data of the sample is all staored in the same  column\n",
    "    to access the data, it needs to be splited in different columns\n",
    "    '''\n",
    "    if len(data.columns) > 10:\n",
    "        Name_sample = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(NAME_FILE)))\n",
    "    else:\n",
    "        print('File', os.path.basename(NAME_FILE), 'contains one sample only.')\n",
    "        Name_sample = data.columns.tolist()[9]\n",
    "    \n",
    "    tmp_df_2 = data[Name_sample].str.split(':', expand=True)\n",
    "    # I have seen very unlikely Na values\n",
    "    # To solve this, they are deleted\n",
    "    tmp_df_2 = tmp_df_2.fillna(value=np.nan)\n",
    "    tmp_df_2 = tmp_df_2.dropna(axis=1, how='any')\n",
    "    # To get the name of the new columns created\n",
    "    tmp_df_2.columns = data.FORMAT.iloc[0].split(':')\n",
    "    return pd.concat([data, tmp_df_2], axis=1),Name_sample\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File W1717693_S1.vcf contains one sample only.\n"
     ]
    }
   ],
   "source": [
    "dataA,Name_sample_1 = take_sample(dataA,NAME_FILE_1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_1(data):\n",
    "    return data.query('FILTER == \"PASS\"')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_2(data):\n",
    "    if \"VF\" in data.columns:\n",
    "        data[\"VF\"] = data[\"VF\"].astype(float)\n",
    "    return data.query('VF > 0.4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataA = filter_2(dataA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def comparason(data):\n",
    "    return dataA[\"CHROM\"].apply(str)+\".\"+dataA[\"POS\"].apply(str)+dataA[\"REF\"]+\".\"+dataA[\"ALT\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-31-61ab00b914a6>:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dataA['CHROMPOSREFALT'] = comparason(dataA)\n"
     ]
    }
   ],
   "source": [
    "dataA['CHROMPOSREFALT'] = comparason(dataA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "frames = [dataA[['CHROMPOSREFALT', 'GT']],dataA[['CHROMPOSREFALT', 'GT']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMPOSREFALT</th>\n",
       "      <th>GT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1.36933096T.C</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr2.25463483G.A</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr2.25466888G.T</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr2.25469502C.T</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr2.198266834T.C</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>chrX.76940534A.G</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>chrX.129147079T.C</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chrX.129147373G.A</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>chrX.129186012C.T</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>chrX.133511988G.A</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>146 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROMPOSREFALT   GT\n",
       "10    chr1.36933096T.C  0/1\n",
       "16    chr2.25463483G.A  0/1\n",
       "19    chr2.25466888G.T  1/1\n",
       "30    chr2.25469502C.T  0/1\n",
       "33   chr2.198266834T.C  0/1\n",
       "..                 ...  ...\n",
       "391   chrX.76940534A.G  1/1\n",
       "423  chrX.129147079T.C  1/1\n",
       "424  chrX.129147373G.A  1/1\n",
       "439  chrX.129186012C.T  1/1\n",
       "447  chrX.133511988G.A  1/1\n",
       "\n",
       "[146 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "semi_final_df = pd.concat(frames)\n",
    "semi_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To see is lists are storing data correctly\n",
    "print(sampleA)\n",
    "print(sampleB)\n",
    "print(Total_positions_comp)\n",
    "print(Pos_same_genotype)\n",
    "print(Pos_different_genotyope)\n",
    "print(Hom)\n",
    "print(Het)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save lists in a df\n",
    "results = pd.DataFrame(list(zip(sampleA, sampleB,Total_positions_comp,Pos_same_genotype,Pos_different_genotyope,Hom,Het)),\n",
    "               columns=['Sample_A','Sample_B','Total_positions_comp','Pos_same_genotype','Pos_different_genotyope','Hom','Het'])\n",
    "results = results[1:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To get proportion of sites with common genotype\n",
    "results[\"identical_position (%)\"] = results[\"Pos_same_genotype\"]/results[\"Total_positions_comp\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To save \n",
    "total_results = pd.read_csv(\"/Users/monkiky/Desktop/VCF-matcher/Samples/Same_patients/Results/both_patients.csv\")\n",
    "total_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To generater the plot\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_context('notebook')\n",
    "sns.scatterplot(data=total_results, x=\"identical_position (%)\", y=\"Total_positions_comp\",\n",
    "               hue = \"Samples:\",\n",
    "               palette=['red',\"dodgerblue\"], legend='full',\n",
    "               ).set(xlabel=\"Proportion of positions with common genotypes\",\n",
    "                    ylabel=\"Number of positions compared\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "################### NOOO DELETE THIS ######################################\n",
    "# A Dashboard to show results in case I will needed\n",
    "\n",
    "\n",
    "### 6. Generate a report in a dashboard\n",
    "# Bootstrap themes by Ann: https://hellodash.pythonanywhere.com/theme_explorer\n",
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.SIMPLEX])\n",
    "\n",
    "app.layout = dbc.Container([\n",
    "    dbc.Row([\n",
    "        dbc.Col([\n",
    "            dbc.Card([\n",
    "                dbc.CardImg(src='./assets/wrgllogohighres.jpeg')\n",
    "                ],color=\"light\", outline=True),\n",
    "                ],width=3),\n",
    "                dbc.Col([\n",
    "                    dbc.Card([\n",
    "                        dbc.CardImg(src='./assets/salisbury-nhs-foundation-trustLeft.png')\n",
    "                             ],color=\"light\", outline=True)\n",
    "                        ],width=2)\n",
    "            ],justify=\"between\"),\n",
    "        dbc.Row(\n",
    "            [dbc.Card([ \n",
    "                dbc.CardBody([\n",
    "                    html.Div([\n",
    "                    html.H1(\"Report: Do the samples belong to the same patient?\"),\n",
    "                    html.H3(\"This application takes common variants and compares the GT fields. A relative large number of common variants which GT fields match is indicative that both samples have a close relationship.\"),\n",
    "                ], style={'textAlign': 'center'}) \n",
    "                     ], style={'textLeft':'center', })      \n",
    "                      ], style={'height':'10vh'},color=\"light\", outline=True),\n",
    "        ]),\n",
    "        dbc.Row([\n",
    "            dbc.Card([ \n",
    "                dbc.CardBody([\n",
    "                    html.Div([\n",
    "                    html.H1(\"\"),\n",
    "                    html.H3(\"\"),\n",
    "                ], style={'textAlign': 'center'}) \n",
    "                     ], style={'textLeft':'center', })      \n",
    "                      ], style={'height':'10vh'},color=\"light\", outline=True),\n",
    "        ]),\n",
    "        dbc.Row([  \n",
    "            dbc.Col([\n",
    "            dash_table.DataTable(\n",
    "        id='datatable-interactivity',\n",
    "        columns=[\n",
    "            {\"name\": i, \"id\": i, \"deletable\": True, \"selectable\": True, \"hideable\": True}\n",
    "            if i == \"iso_alpha3\" or i == \"year\" or i == \"id\"\n",
    "            else {\"name\": i, \"id\": i, \"deletable\": True, \"selectable\": True}\n",
    "            for i in final_df.columns\n",
    "        ],\n",
    "        data=final_df.to_dict('records'),  # the contents of the table html.H3(os.path.basename(Name_file1)),\n",
    "        filter_action=\"native\",     # allow filtering of data by user ('native') or not ('none')\n",
    "        sort_action=\"native\",       # enables data to be sorted per-column by user or not ('none')\n",
    "        sort_mode=\"single\",         # sort across 'multi' or 'single' columns\n",
    "        row_deletable=True,         # choose if user can delete a row (True) or not (False)\n",
    "        selected_rows=[],           # indices of rows that user selects\n",
    "        page_action=\"native\",       # all data is passed to the table up-front or not ('none')\n",
    "        page_current=0,             # page number that user is on\n",
    "        style_cell={                # ensure adequate header width when text is shorter than cell's text\n",
    "            'minWidth': 200, 'maxWidth': 95, 'width': 95\n",
    "        },\n",
    "        style_data_conditional=[\n",
    "        {\n",
    "            'if': {'row_index': 'odd'},\n",
    "            'backgroundColor': 'rgb(248, 248, 248)'\n",
    "        }\n",
    "    ],\n",
    "        style_header={\n",
    "        'backgroundColor': 'rgb(230, 230, 230)',\n",
    "        'fontWeight': 'bold'\n",
    "    })]),\n",
    "    dbc.Col([\n",
    "        dbc.Card([ \n",
    "            dbc.CardBody([\n",
    "                html.Div([\n",
    "                    html.H4(\"Number of common variants:\",style={'color': 'black', 'fontSize': 15}),\n",
    "                    html.H4(len(final_df)),\n",
    "                    html.H4(\"Number of matches:\", style={'color': 'black', 'fontSize': 15}),\n",
    "                    html.H4(final_df.Matches.value_counts()[1]),\n",
    "                    ], \n",
    "                    style={'textAlign': 'left'})\n",
    "                    ])\n",
    "                    ],\n",
    "                    style={'height':'8vh'},color=\"light\", outline=True),\n",
    "                    ])\n",
    "                    ])\n",
    "                    ])\n",
    "\n",
    "\n",
    "\n",
    "  \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It seems that test_sample.vcf has more than one sample. Please introduce the name of the sample you wish to analyse:  373978824\n",
      "It seems that test_sample.vcf has more than one sample. Please introduce the name of the sample you wish to analyse:  373978833\n",
      "\n",
      " _____________________________  REPORT  ________________________________________ \n",
      "\n",
      "vcf 1: test_sample.vcf  AND its sample name: 373978824  \n",
      "vcf 2:  test_sample.vcf  AND its sample name: 373978833\n",
      "\n",
      "                                                  Homozygous: 954 \n",
      "Number of positions with the same genotype: 1012 \n",
      "                                                  Heterozygous: 58 \n",
      "                                                \n",
      "                                                  \n",
      "Number of positions with different genotype: 330 \n",
      "                                                  \n",
      "\n",
      "\n",
      "Total positions compared: 1342\n",
      "Percentage in common: 1012/1342= 0.7540983606557377\n",
      "\n",
      " ____________________________ END REPORT  _______________________________________\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "'''\n",
    "Created on 07/09/2021\n",
    "\n",
    "@author: manuel.dominguezbecerra@nhs.net\n",
    "\n",
    "'''\n",
    "\n",
    "# Import libraries\n",
    "import os\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# To allow the user to introduce the file after running \"python run.ppy\"\n",
    "\n",
    "# Example:    python run.py --vcf1 /path/file_name1 --vcf2 /path/file_name2\n",
    "\n",
    "   ### 1. Load the files\n",
    "\n",
    "# Take only the data contains in the body of the vcf files.\n",
    "\n",
    "def load_sample (Name_file):\n",
    "    '''\n",
    "    Take the header of the body of the CSV file\n",
    "    '''\n",
    "    with open(Name_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#') and len(line)>2 and line[1] != '#':\n",
    "                columns = line[1:-1].split('\\t')\n",
    "                data = pd.read_csv(Name_file, comment='#', delimiter='\\t', names=columns)\n",
    "                break\n",
    "    return data\n",
    "\n",
    "    ### 2. One or > one sample in the files? \n",
    "\n",
    "def take_sample(data,NAME_FILE):\n",
    "    '''\n",
    "    VCF files can have 1 or > 1 samples\n",
    "    We need to know what sample the user wants to compare\n",
    "    After this, the data of the sample is all stored in one  column\n",
    "    to access the data, it needs to  split it in different columns\n",
    "    '''\n",
    "    if len(data.columns) > 10:\n",
    "        Name_sample = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(NAME_FILE)))\n",
    "    else:\n",
    "        print('File', os.path.basename(NAME_FILE), 'contains one sample only.')\n",
    "        Name_sample = data.columns.tolist()[9]\n",
    "    \n",
    "    tmp_df_2 = data[Name_sample].str.split(':', expand=True)\n",
    "    # I have seen very unlikely Na values\n",
    "    # To solve this, they are deleted\n",
    "    tmp_df_2 = tmp_df_2.fillna(value=np.nan)\n",
    "    tmp_df_2 = tmp_df_2.dropna(axis=1, how='any')\n",
    "    # To get the name of the new columns created\n",
    "    tmp_df_2.columns = data.FORMAT.iloc[0].split(':')\n",
    "    return pd.concat([data, tmp_df_2], axis=1),Name_sample\n",
    "    \n",
    "    ### 3. Filter the variants\n",
    "\n",
    "def filter_1(data):\n",
    "    '''\n",
    "    Take only the variants that \"PASS\" the \"FILTER\" field\n",
    "    '''\n",
    "    return data.query('FILTER == \"PASS\"')\n",
    "\n",
    "\n",
    "def filter_2(data):\n",
    "    '''\n",
    "    Take the variants which VF > 0.4\n",
    "    '''\n",
    "    if \"VF\" in data.columns:\n",
    "        data[\"VF\"] = data[\"VF\"].astype(float)\n",
    "        return data.query('VF > 0.4')\n",
    "    else:\n",
    "        return data\n",
    "\n",
    "   ### 4. Variants comparison\n",
    "\n",
    "def Unit_identifier(data):\n",
    "    '''\n",
    "    Two variants are the same if  CHROM, POS, REF, ALT and GT are equal\n",
    "    Here, the unit identifier is created\n",
    "    '''\n",
    "    return data[\"CHROM\"].apply(str)+\".\"+data[\"POS\"].apply(str)+data[\"REF\"]+\".\"+data[\"ALT\"]\n",
    "\n",
    "\n",
    "def concatenate_samples(data1,data2):\n",
    "    '''\n",
    "    To proceeded, both data frames need to be concatenate \n",
    "    '''\n",
    "    frames = [data1[['CHROMPOSREFALT', 'GT']],data2[['CHROMPOSREFALT', 'GT']]]\n",
    "    return pd.concat(frames)\n",
    "\n",
    "def match_variants(data):\n",
    "    '''\n",
    "    Match common variants in one column and put the GT values of both samples\n",
    "    in different columns.\n",
    "    No common variants are also set in the first column giving NaN values \n",
    "    in the third column. I have solved this by substituting the NaN for 99/99 values\n",
    "    '''\n",
    "    matched_df = (data.assign(key=data.groupby('CHROMPOSREFALT').cumcount())\n",
    "        .pivot('CHROMPOSREFALT','key','GT')\n",
    "        .rename(columns=lambda x:f\"Sample{x+1}\")\n",
    "        .rename_axis(columns=None).reset_index())\n",
    "    matched_df = matched_df.replace(np.NaN,\"99/99\" )\n",
    "    return matched_df.iloc[:, : 3]\n",
    "\n",
    "def count_hom_het_variants(data):\n",
    "    data['Genotype1'] = data.iloc[:,2].apply(lambda x: x.split('/' or '|')[0])\n",
    "    data['Genotype2'] = data.iloc[:,2].apply(lambda x: x.split('/' or '|')[1])\n",
    "    return np.select([data['Matches'] & data['Genotype1'].eq(data['Genotype2']),\n",
    "                       data['Matches'] & data['Genotype1'].ne(data['Genotype2'])],\n",
    "                       choicelist=[\"Hom\", \"Het\"],\n",
    "                       default=pd.NA)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # execute only if run as a script\n",
    "\n",
    "    # Enter the path/file names\n",
    "\n",
    "    NAME_FILE_1 = \"/Users/monkiky/Desktop/VCF_matcher/Testing/test_sample.vcf\"\n",
    "    NAME_FILE_2 = \"/Users/monkiky/Desktop/VCF_matcher/Testing/test_sample.vcf\"\n",
    "\n",
    "    # Load the files\n",
    "    dataA = load_sample(NAME_FILE_1)\n",
    "    dataB = load_sample(NAME_FILE_2)\n",
    "\n",
    "    # Save now the data of this sample and its name (we will need for the report)\n",
    "    dataA,Name_sample_1 = take_sample(dataA,NAME_FILE_1)\n",
    "    dataB,Name_sample_2 = take_sample(dataB,NAME_FILE_2)\n",
    "\n",
    "    # Filter 1\n",
    "    dataA = filter_1(dataA)\n",
    "    dataB = filter_1(dataB)\n",
    "\n",
    "    # Filter 2\n",
    "    dataA = filter_2(dataA)\n",
    "    dataB = filter_2(dataB)\n",
    "\n",
    "    # Unit_identifier\n",
    "    dataA['CHROMPOSREFALT'] = Unit_identifier(dataA)\n",
    "    dataB['CHROMPOSREFALT'] = Unit_identifier(dataB)\n",
    "\n",
    "\n",
    "    # Concatenate both data frames\n",
    "    semi_final_df = concatenate_samples(dataA,dataB)\n",
    "\n",
    "    # Match common variants\n",
    "    final_df= match_variants(semi_final_df)\n",
    "\n",
    "#### Generate report ####\n",
    "\n",
    "    # Variables needed for both type of reports\n",
    "\n",
    "    r0 = os.path.basename(NAME_FILE_1) # Get the  file name of a path/file_name\n",
    "    r1 = Name_sample_1                 # Get the name of the sample 1\n",
    "    r2 = os.path.basename(NAME_FILE_2)\n",
    "    r3 = Name_sample_2\n",
    "\n",
    "    if \"Sample2\" not in final_df.columns:\n",
    "    # If Sample2 is not in final_df, that means there is not common position between the samples.\n",
    "    # This is a the most simple kind of report with only the name of the files\n",
    "    # and samples and informing that no common positions have been found between the samples\n",
    "        REPORT_A = '''\n",
    " _____________________________  REPORT  ________________________________________ \n",
    "\n",
    "   vcf 1: {0}  AND its sample name: {1}  \n",
    "   vcf 2:  {2}  AND its sample name: {3}\n",
    "\n",
    "   No common positions found between samples\n",
    "\n",
    " ____________________________ END REPORT  _______________________________________\n",
    "'''\n",
    "        print(REPORT_A.format(r0,r1,r2,r3,))\n",
    "\n",
    "    # If Sample2 is in final_df.columns, that means that at least one common variant has been found\n",
    "    # so, generate the data from these common variants\n",
    "    else:\n",
    "    # Else count common positions and carry on the report.\n",
    "        final_df[\"Matches\"] = np.where(final_df[\"Sample1\"] == final_df[\"Sample2\"], True, False)\n",
    "        final_df.columns = ['CHROMPOSREFALT',os.path.basename(NAME_FILE_1),os.path.basename(NAME_FILE_2),\"Matches\" ]\n",
    "\n",
    "    #  Get and count hom and het common variants\n",
    "        final_df['Hom/het'] = count_hom_het_variants(final_df)\n",
    "\n",
    "    # Generate the second type of report\n",
    "\n",
    "        r4 = final_df['Matches'].value_counts().get(True, 0) # Count trues if any, return 0\n",
    "        r5 = final_df['Hom/het'].value_counts().get(\"Hom\", 0)\n",
    "        r6 =final_df['Hom/het'].value_counts().get(\"Het\", 0)\n",
    "        r7 = final_df['Matches'].value_counts().get(False, 0)\n",
    "        r8 = len(final_df)\n",
    "        r9 = r4/(r4+r7)\n",
    "        REPORT_B = '''\n",
    " _____________________________  REPORT  ________________________________________ \n",
    "\n",
    "vcf 1: {0}  AND its sample name: {1}  \n",
    "vcf 2:  {2}  AND its sample name: {3}\n",
    "\n",
    "                                                  Homozygous: {5} \n",
    "Number of positions with the same genotype: {4} \n",
    "                                                  Heterozygous: {6} \n",
    "                                                \n",
    "                                                  \n",
    "Number of positions with different genotype: {7} \n",
    "                                                  \n",
    "\n",
    "\n",
    "Total positions compared: {8}\n",
    "Percentage in common: {4}/{8}= {9}\n",
    "\n",
    " ____________________________ END REPORT  _______________________________________\n",
    "'''\n",
    "        print(REPORT_B.format(r0,r1,r2,r3,r4,r5,r6,r7,r8,r9))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>1113412920</th>\n",
       "      <th>...</th>\n",
       "      <th>373978800</th>\n",
       "      <th>373978809</th>\n",
       "      <th>373978824</th>\n",
       "      <th>373978833</th>\n",
       "      <th>GT</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>GQ</th>\n",
       "      <th>PL</th>\n",
       "      <th>CHROMPOSREFALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2160224</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>10.30</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;AF=0.059;AN=34;DP=41;ExcessHet=0.0663;FS=...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:0,0:0:.:0,0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:2,0:2:6:0,6,49</td>\n",
       "      <td>0/0:2,0:2:6:0,6,49</td>\n",
       "      <td>0/0:2,0:2:6:0,6,49</td>\n",
       "      <td>0/0:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0,0</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>1.2160224G.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2160661</td>\n",
       "      <td>rs149898447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>947.71</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.021;AN=48;BaseQRankSum=-0.58;Clippin...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:68,0:68:99:0,120,1800</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:71,0:71:99:0,120,1800</td>\n",
       "      <td>0/0:18,0:18:48:0,48,720</td>\n",
       "      <td>0/0:58,0:58:99:0,120,1800</td>\n",
       "      <td>0/0:39,0:39:99:0,102,1530</td>\n",
       "      <td>0/0</td>\n",
       "      <td>39,0</td>\n",
       "      <td>39</td>\n",
       "      <td>99</td>\n",
       "      <td>0,102,1530</td>\n",
       "      <td>1.2160661C.T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2161003</td>\n",
       "      <td>rs149642284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>667.71</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.021;AN=48;BaseQRankSum=-1.926;Clippi...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:68,0:68:99:0,120,1800</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:55,0:55:99:0,120,1800</td>\n",
       "      <td>0/0:36,0:36:99:0,108,905</td>\n",
       "      <td>0/0:56,0:56:99:0,120,1800</td>\n",
       "      <td>0/0:57,0:57:99:0,120,1800</td>\n",
       "      <td>0/0</td>\n",
       "      <td>57,0</td>\n",
       "      <td>57</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "      <td>1.2161003C.T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2234488</td>\n",
       "      <td>rs140745811</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1743.71</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.021;AN=48;BaseQRankSum=1.24;Clipping...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:220,0:220:99:0,120,1800</td>\n",
       "      <td>...</td>\n",
       "      <td>0/1:123,90:213:99:1781,0,2543</td>\n",
       "      <td>0/0:110,0:110:99:0,120,1800</td>\n",
       "      <td>0/0:191,0:191:99:0,120,1800</td>\n",
       "      <td>0/0:177,0:177:99:0,120,1800</td>\n",
       "      <td>0/0</td>\n",
       "      <td>177,0</td>\n",
       "      <td>177</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "      <td>1.2234488C.T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2235378</td>\n",
       "      <td>rs140889128</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>187.18</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=3;AF=0.063;AN=48;BaseQRankSum=0.674;Clippin...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:2,0:2:6:0,6,49</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:4,0:4:12:0,12,99</td>\n",
       "      <td>0/0:6,0:6:18:0,18,148</td>\n",
       "      <td>0/1:3,3:6:56:61,0,56</td>\n",
       "      <td>0/0:1,0:1:3:0,3,42</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0,3,42</td>\n",
       "      <td>1.2235378C.T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>X</td>\n",
       "      <td>153588893</td>\n",
       "      <td>rs79391751</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>417.55</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;AF=0.042;AN=48;BaseQRankSum=-0.074;Clippi...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:2,0:2:3:0,3,45</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:14,0:14:36:0,36,540</td>\n",
       "      <td>0/0:7,0:7:18:0,18,270</td>\n",
       "      <td>0/0:7,0:7:15:0,15,225</td>\n",
       "      <td>0/0:7,0:7:12:0,12,180</td>\n",
       "      <td>0/0</td>\n",
       "      <td>7,0</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>0,12,180</td>\n",
       "      <td>X.153588893G.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>X</td>\n",
       "      <td>153591022</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1475.99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;AF=0.042;AN=48;DP=1582;ExcessHet=0.0464;F...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:0,56:56:99:1530,167,0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:33,0:33:93:0,93,1395</td>\n",
       "      <td>0/0:28,0:28:84:0,84,736</td>\n",
       "      <td>0/0:110,1:111:99:0,120,1800</td>\n",
       "      <td>0/0:45,0:45:99:0,120,1800</td>\n",
       "      <td>0/0</td>\n",
       "      <td>45,0</td>\n",
       "      <td>45</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "      <td>X.153591022G.T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>X</td>\n",
       "      <td>153592948</td>\n",
       "      <td>rs73638274</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2729.71</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.021;AN=48;BaseQRankSum=-3.283;Clippi...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:118,0:118:99:0,120,1800</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:118,1:119:99:0,120,1800</td>\n",
       "      <td>0/0:54,0:54:99:0,120,1800</td>\n",
       "      <td>0/0:192,0:192:99:0,120,1800</td>\n",
       "      <td>0/0:84,0:84:99:0,120,1800</td>\n",
       "      <td>0/0</td>\n",
       "      <td>84,0</td>\n",
       "      <td>84</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "      <td>X.153592948G.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1398</th>\n",
       "      <td>X</td>\n",
       "      <td>153594535</td>\n",
       "      <td>rs36051194</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>227.71</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;AF=0.021;AN=48;BaseQRankSum=0;ClippingRan...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:19,0:19:51:0,51,765</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:22,0:22:60:0,60,900</td>\n",
       "      <td>0/0:14,0:14:27:0,27,405</td>\n",
       "      <td>0/0:43,0:43:99:0,120,1800</td>\n",
       "      <td>0/0:27,0:27:75:0,75,1125</td>\n",
       "      <td>0/0</td>\n",
       "      <td>27,0</td>\n",
       "      <td>27</td>\n",
       "      <td>75</td>\n",
       "      <td>0,75,1125</td>\n",
       "      <td>X.153594535G.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>X</td>\n",
       "      <td>153595865</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2583.99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;AF=0.042;AN=48;BaseQRankSum=-2.598;Clippi...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/0:190,0:190:99:0,120,1800</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:156,0:156:99:0,120,1800</td>\n",
       "      <td>0/0:91,0:91:99:0,120,1800</td>\n",
       "      <td>0/0:298,0:298:99:0,120,1800</td>\n",
       "      <td>1/1:1,104:105:99:2638,272,0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1,104</td>\n",
       "      <td>105</td>\n",
       "      <td>99</td>\n",
       "      <td>2638,272,0</td>\n",
       "      <td>X.153595865A.G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1352 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS           ID REF ALT     QUAL FILTER  \\\n",
       "0        1    2160224            .   G   A    10.30   PASS   \n",
       "1        1    2160661  rs149898447   C   T   947.71   PASS   \n",
       "2        1    2161003  rs149642284   C   T   667.71   PASS   \n",
       "3        1    2234488  rs140745811   C   T  1743.71   PASS   \n",
       "4        1    2235378  rs140889128   C   T   187.18   PASS   \n",
       "...    ...        ...          ...  ..  ..      ...    ...   \n",
       "1394     X  153588893   rs79391751   G   A   417.55   PASS   \n",
       "1395     X  153591022            .   G   T  1475.99   PASS   \n",
       "1396     X  153592948   rs73638274   G   A  2729.71   PASS   \n",
       "1398     X  153594535   rs36051194   G   A   227.71   PASS   \n",
       "1399     X  153595865            .   A   G  2583.99   PASS   \n",
       "\n",
       "                                                   INFO          FORMAT  \\\n",
       "0     AC=2;AF=0.059;AN=34;DP=41;ExcessHet=0.0663;FS=...  GT:AD:DP:GQ:PL   \n",
       "1     AC=1;AF=0.021;AN=48;BaseQRankSum=-0.58;Clippin...  GT:AD:DP:GQ:PL   \n",
       "2     AC=1;AF=0.021;AN=48;BaseQRankSum=-1.926;Clippi...  GT:AD:DP:GQ:PL   \n",
       "3     AC=1;AF=0.021;AN=48;BaseQRankSum=1.24;Clipping...  GT:AD:DP:GQ:PL   \n",
       "4     AC=3;AF=0.063;AN=48;BaseQRankSum=0.674;Clippin...  GT:AD:DP:GQ:PL   \n",
       "...                                                 ...             ...   \n",
       "1394  AC=2;AF=0.042;AN=48;BaseQRankSum=-0.074;Clippi...  GT:AD:DP:GQ:PL   \n",
       "1395  AC=2;AF=0.042;AN=48;DP=1582;ExcessHet=0.0464;F...  GT:AD:DP:GQ:PL   \n",
       "1396  AC=1;AF=0.021;AN=48;BaseQRankSum=-3.283;Clippi...  GT:AD:DP:GQ:PL   \n",
       "1398  AC=1;AF=0.021;AN=48;BaseQRankSum=0;ClippingRan...  GT:AD:DP:GQ:PL   \n",
       "1399  AC=2;AF=0.042;AN=48;BaseQRankSum=-2.598;Clippi...  GT:AD:DP:GQ:PL   \n",
       "\n",
       "                       1113412920  ...                      373978800  \\\n",
       "0               0/0:0,0:0:.:0,0,0  ...             0/0:2,0:2:6:0,6,49   \n",
       "1       0/0:68,0:68:99:0,120,1800  ...      0/0:71,0:71:99:0,120,1800   \n",
       "2       0/0:68,0:68:99:0,120,1800  ...      0/0:55,0:55:99:0,120,1800   \n",
       "3     0/0:220,0:220:99:0,120,1800  ...  0/1:123,90:213:99:1781,0,2543   \n",
       "4              0/0:2,0:2:6:0,6,49  ...           0/0:4,0:4:12:0,12,99   \n",
       "...                           ...  ...                            ...   \n",
       "1394           0/0:2,0:2:3:0,3,45  ...        0/0:14,0:14:36:0,36,540   \n",
       "1395    1/1:0,56:56:99:1530,167,0  ...       0/0:33,0:33:93:0,93,1395   \n",
       "1396  0/0:118,0:118:99:0,120,1800  ...    0/0:118,1:119:99:0,120,1800   \n",
       "1398      0/0:19,0:19:51:0,51,765  ...        0/0:22,0:22:60:0,60,900   \n",
       "1399  0/0:190,0:190:99:0,120,1800  ...    0/0:156,0:156:99:0,120,1800   \n",
       "\n",
       "                        373978809                    373978824  \\\n",
       "0              0/0:2,0:2:6:0,6,49           0/0:2,0:2:6:0,6,49   \n",
       "1         0/0:18,0:18:48:0,48,720    0/0:58,0:58:99:0,120,1800   \n",
       "2        0/0:36,0:36:99:0,108,905    0/0:56,0:56:99:0,120,1800   \n",
       "3     0/0:110,0:110:99:0,120,1800  0/0:191,0:191:99:0,120,1800   \n",
       "4           0/0:6,0:6:18:0,18,148         0/1:3,3:6:56:61,0,56   \n",
       "...                           ...                          ...   \n",
       "1394        0/0:7,0:7:18:0,18,270        0/0:7,0:7:15:0,15,225   \n",
       "1395      0/0:28,0:28:84:0,84,736  0/0:110,1:111:99:0,120,1800   \n",
       "1396    0/0:54,0:54:99:0,120,1800  0/0:192,0:192:99:0,120,1800   \n",
       "1398      0/0:14,0:14:27:0,27,405    0/0:43,0:43:99:0,120,1800   \n",
       "1399    0/0:91,0:91:99:0,120,1800  0/0:298,0:298:99:0,120,1800   \n",
       "\n",
       "                        373978833   GT     AD   DP  GQ          PL  \\\n",
       "0               0/0:0,0:0:.:0,0,0  0/0    0,0    0   .       0,0,0   \n",
       "1       0/0:39,0:39:99:0,102,1530  0/0   39,0   39  99  0,102,1530   \n",
       "2       0/0:57,0:57:99:0,120,1800  0/0   57,0   57  99  0,120,1800   \n",
       "3     0/0:177,0:177:99:0,120,1800  0/0  177,0  177  99  0,120,1800   \n",
       "4              0/0:1,0:1:3:0,3,42  0/0    1,0    1   3      0,3,42   \n",
       "...                           ...  ...    ...  ...  ..         ...   \n",
       "1394        0/0:7,0:7:12:0,12,180  0/0    7,0    7  12    0,12,180   \n",
       "1395    0/0:45,0:45:99:0,120,1800  0/0   45,0   45  99  0,120,1800   \n",
       "1396    0/0:84,0:84:99:0,120,1800  0/0   84,0   84  99  0,120,1800   \n",
       "1398     0/0:27,0:27:75:0,75,1125  0/0   27,0   27  75   0,75,1125   \n",
       "1399  1/1:1,104:105:99:2638,272,0  1/1  1,104  105  99  2638,272,0   \n",
       "\n",
       "      CHROMPOSREFALT  \n",
       "0       1.2160224G.A  \n",
       "1       1.2160661C.T  \n",
       "2       1.2161003C.T  \n",
       "3       1.2234488C.T  \n",
       "4       1.2235378C.T  \n",
       "...              ...  \n",
       "1394  X.153588893G.A  \n",
       "1395  X.153591022G.T  \n",
       "1396  X.153592948G.A  \n",
       "1398  X.153594535G.A  \n",
       "1399  X.153595865A.G  \n",
       "\n",
       "[1352 rows x 39 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMPOSREFALT</th>\n",
       "      <th>GT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.2160224G.A</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.2160661C.T</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.2161003C.T</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.2234488C.T</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.2235378C.T</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>X.153588893G.A</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>X.153591022G.T</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>X.153592948G.A</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1398</th>\n",
       "      <td>X.153594535G.A</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>X.153595865A.G</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROMPOSREFALT   GT\n",
       "0       1.2160224G.A  0/0\n",
       "1       1.2160661C.T  0/0\n",
       "2       1.2161003C.T  0/0\n",
       "3       1.2234488C.T  0/0\n",
       "4       1.2235378C.T  0/1\n",
       "...              ...  ...\n",
       "1394  X.153588893G.A  0/0\n",
       "1395  X.153591022G.T  0/0\n",
       "1396  X.153592948G.A  0/0\n",
       "1398  X.153594535G.A  0/0\n",
       "1399  X.153595865A.G  1/1\n",
       "\n",
       "[2704 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "semi_final_df = concatenate_samples(dataA,dataB)\n",
    "semi_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMPOSREFALT</th>\n",
       "      <th>Sample1</th>\n",
       "      <th>Sample2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.12009955C.T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.12009956G.A</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.12010469G.T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.12017959A.G</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.12024235C.T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1337</th>\n",
       "      <td>X.595475C.A</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1338</th>\n",
       "      <td>X.8503641G.A</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1339</th>\n",
       "      <td>X.8504833C.T</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1340</th>\n",
       "      <td>X.9707566C.T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1341</th>\n",
       "      <td>X.9711595G.C</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1342 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROMPOSREFALT Sample1 Sample2\n",
       "0     1.12009955C.T     0/0     0/0\n",
       "1     1.12009956G.A     0/0     0/0\n",
       "2     1.12010469G.T     0/0     0/0\n",
       "3     1.12017959A.G     0/0     0/0\n",
       "4     1.12024235C.T     0/0     0/0\n",
       "...             ...     ...     ...\n",
       "1337    X.595475C.A     0/0     0/0\n",
       "1338   X.8503641G.A     1/1     1/1\n",
       "1339   X.8504833C.T     1/1     1/1\n",
       "1340   X.9707566C.T     0/0     0/0\n",
       "1341   X.9711595G.C     0/0     0/0\n",
       "\n",
       "[1342 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df= match_variants(semi_final_df)\n",
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
