{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48c0907f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File W2008872_S8-J4PJL.vcf contains one sample only.\n",
      "It seems that 210513_M00321_0553_000000000-JK4CY_Filtered_Annotated.vcf has more than one sample. Please introduce the name of the sample you wish to analyse:  373977987\n",
      "No common position found in both samples\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "'''\n",
    "Created on 09/12/2014\n",
    "\n",
    "@author: manuel.dominguezbecerra@nhs.net\n",
    "\n",
    "'''\n",
    "\n",
    "\n",
    "# Import libraries \n",
    "import argparse\n",
    "import pandas as pd                      # pip install dash\n",
    "import os\n",
    "import subprocess\n",
    "import numpy as np\n",
    "\n",
    "import dash                              # pip install dash\n",
    "import dash_html_components as html\n",
    "import dash_core_components as dcc\n",
    "from dash.dependencies import Output, Input\n",
    "\n",
    "from dash_extensions import Lottie       # pip install dash-extensions\n",
    "import dash_bootstrap_components as dbc  # pip install dash-bootstrap-components\n",
    "import plotly.express as px              # pip install plotly\n",
    "from datetime import date\n",
    "import calendar\n",
    "from wordcloud import WordCloud          # pip install wordcloud\n",
    "import sqlite3                           # pip install sqlite3\n",
    "import plotly.graph_objects as go\n",
    "import dash_table\n",
    "from dash.dependencies import Input, Output, State\n",
    "\n",
    "\n",
    "# Take the samples path/name_file\n",
    "\n",
    "#parser = argparse.ArgumentParser()\n",
    "#parser.add_argument('--vcf1', type=str, required=True)\n",
    "#parser.add_argument('--vcf2', type=int, required=True)\n",
    "#args = parser.parse_args()\n",
    "\n",
    "#Name_file1 = args.vcf1 # Remove comments when finished\n",
    "#Name_file2 = args.vcf2\n",
    "\n",
    "## Delete this when finished\n",
    "#Name_file1 = \"/Users/monkiky/Desktop/Do-these-samples-belong-to-the-same-patient/Samples/W2008872_S8-J4PJL_copy.vcf\"\n",
    "Name_file1 = \"/Users/monkiky/Desktop/VCF-matcher/Samples/W2008872_S8-J4PJL.vcf\"\n",
    "#Name_file1 = \"/Users/monkiky/Desktop/Do-these-samples-belong-to-the-same-patient/Samples/W2103014_S10-JGFJG.vcf\"\n",
    "Name_file2 = \"/Users/monkiky/Desktop/VCF-matcher/Samples/Myeloid_1.2/210513_M00321_0553_000000000-JK4CY_Filtered_Annotated.vcf\"\n",
    "\n",
    "\n",
    "### 1. Remove the header of input files\n",
    "\n",
    "# File A\n",
    "cmd = \"sed '/^##/ d'  {0} >  FileA.txt\".format(Name_file1)\n",
    "os.system(cmd)\n",
    "\n",
    "# File B \n",
    "cmd = \"sed '/^##/ d'  {0} >  FileB.txt\".format(Name_file2)\n",
    "os.system(cmd)\n",
    "\n",
    "# Files to pandas object\n",
    "dataA = pd.read_csv(\"./FileA.txt\", delimiter = \"\\t\" )\n",
    "dataB = pd.read_csv(\"./FileB.txt\", delimiter = \"\\t\" )\n",
    "\n",
    "# Delete files created\n",
    "cmd = os. getcwd()+\"/FileA.txt\"\n",
    "os.remove(cmd)\n",
    "\n",
    "cmd = os. getcwd()+\"/FileB.txt\"\n",
    "os.remove(cmd)\n",
    "\n",
    "# If the pd dataframes have more than 10 columns\n",
    "# that means there are more than 1 sample in the vcf file\n",
    "\n",
    "if len(dataA.columns) > 10:\n",
    "     Name_sample1 = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(Name_file1)))\n",
    "else:\n",
    "    print('File', os.path.basename(Name_file1), 'contains one sample only.')\n",
    "    Name_sample1 = dataA.columns.tolist()[9]\n",
    "\n",
    "if len(dataB.columns) > 10:\n",
    "     Name_sample2 = input('It seems that {} has more than one sample. Please introduce the name of the sample you wish to analyse:  '.format(os.path.basename(Name_file2)))\n",
    "else:\n",
    "    print('File', os.path.basename(Name_file2), 'contains one sample only.')\n",
    "    Name_sample2 = dataB.columns.tolist()[9]\n",
    "\n",
    "    ### 2. Filter the variants  FILTER = PASS and Variant Frecuency =>0.4\n",
    "\n",
    "# VF is in the 9th column, we need to extract and put VF data in an independent column first\n",
    "df2 = dataA[Name_sample1].str.split(':', expand=True)\n",
    "\n",
    "df2 = df2.fillna(value=np.nan)\n",
    "df2 = df2.dropna(axis=1, how='any')\n",
    "\n",
    "df2.columns = dataA.FORMAT.iloc[0].split(':')\n",
    "dataA = pd.concat([dataA, df2], axis=1)\n",
    "\n",
    "#df2.columns[df2.columns.str.contains(pat = 'VF')] \n",
    "\n",
    "\n",
    "\n",
    "# Same for dataB\n",
    "df3 = dataB[Name_sample2].str.split(':', expand=True)\n",
    "\n",
    "df3 = df3.fillna(value=np.nan) # Sometimes empty columns are created giving an error as there is more columns than columns names\n",
    "df3 = df3.dropna(axis=1, how='any') # This delete the empty columns\n",
    "\n",
    "df3.columns = dataB.FORMAT.iloc[0].split(':')\n",
    "dataB = pd.concat([dataB, df3], axis=1)\n",
    "\n",
    "# Now, letÂ´s filter the variant\n",
    "# We dont want variants that doesnt pass the Filter status\n",
    "# This field is mandatory, no condition is needed because it is applied to any type of VCF file.\n",
    "dataA = dataA.query('FILTER == \"PASS\"')\n",
    "dataB = dataB.query('FILTER == \"PASS\"')\n",
    "                    \n",
    "# Now, I apply Variant frecuency (VF) filter (non-mandatory field)\n",
    "\n",
    "# If there is a VF Genotype fields in the file\n",
    "# Take the variants that are > 0.4\n",
    "if \"VF\" in dataA.columns:\n",
    "    dataA[\"VF\"] = dataA[\"VF\"].astype(float)\n",
    "    dataA = dataA.query('VF > 0.4')\n",
    "    \n",
    "if \"VF\" in dataB.columns:\n",
    "    dataB[\"VF\"] = dataB[\"VF\"].astype(float)\n",
    "    dataB = dataB.query('VF > 0.4')\n",
    "    \n",
    "\n",
    "### 3. Let's create the Ben's Code, this is CHROM+POS+REF+ALT\n",
    "\n",
    "dataA['CHROMPOSREDALT']=dataA[\"#CHROM\"].apply(str)+\".\"+dataA[\"POS\"].apply(str)+dataA[\"REF\"]+\".\"+dataA[\"ALT\"]\n",
    "\n",
    "dataB['CHROMPOSREDALT']=dataB[\"#CHROM\"].apply(str)+\".\"+dataB[\"POS\"].apply(str)+dataB[\"REF\"]+\".\"+dataB[\"ALT\"]\n",
    "\n",
    "\n",
    "\n",
    "### 4. Take common Ben's Codes in the same column and put the GT of each (both  copies) in independent columns\n",
    "df3 = dataA[['CHROMPOSREDALT', 'GT']].copy()\n",
    "df4 = dataB[['CHROMPOSREDALT', 'GT']].copy()\n",
    "frames = [df3,df4]\n",
    "semi_final_df = pd.concat(frames)\n",
    "\n",
    "\n",
    "final_df=(semi_final_df.assign(key=semi_final_df.groupby('CHROMPOSREDALT').cumcount())\n",
    "      .pivot('CHROMPOSREDALT','key','GT')\n",
    "      .rename(columns=lambda x:f\"Sample{x+1}\")\n",
    "      .rename_axis(columns=None).reset_index())\n",
    "\n",
    "\n",
    "\n",
    "### 5. See how many GT match\n",
    "\n",
    "if \"Sample2\" not in final_df.columns:\n",
    "    print(\"No common position found in both samples\")\n",
    "    print(\"No report generated\")\n",
    "else:    \n",
    "    final_df[\"Matches\"] = np.where(final_df[\"Sample1\"] == final_df[\"Sample2\"], True, False)\n",
    "    final_df.columns = ['CHROM.POS.REF.ALT',os.path.basename(Name_file1),os.path.basename(Name_file2),\"Matches\" ]\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "        \n",
    "        \n",
    "    \n",
    "### 6.1 Generate results in the terminal\n",
    "\n",
    "# Variable to introduce in the report:\n",
    "    r0 = os.path.basename(Name_file1) # Get the  file name of a path/file_name \n",
    "    r1 = Name_sample1\n",
    "    r2 = os.path.basename(Name_file2)\n",
    "    r3 = Name_sample2\n",
    "    r4 = final_df['Matches'].value_counts().get(True, 0) # Count trues if any, retunr 0\n",
    "    r5 = final_df['Matches'].value_counts().get(False, 0)\n",
    "    r6 = len(final_df)\n",
    "    r7 = r4/(r4+r5)\n",
    "    r6 == r4+r5\n",
    "    report = '''\n",
    " _____________________________  REPORT  ________________________________________ \n",
    "\n",
    "vcf 1: {0}  AND its sample name: {1}  \n",
    "vcf 2:  {2}  AND its sample name: {3}\n",
    "\n",
    "                                                  Homozigous: \n",
    "Number of positions with the same genotype: {4} \n",
    "                                                  Heterozigous:\n",
    "                                                \n",
    "                                                  Homozigous:\n",
    "Number of positions with diferent genotype: {5} \n",
    "                                                  Heterozigous: \n",
    "\n",
    "\n",
    "Total positions compared: {6}\n",
    "Percentage in common: {4}/{6}= {7}\n",
    " ____________________________ END REPORT  _______________________________________\n",
    "'''\n",
    "    print(report.format(r0,r1,r2,r3,r4,r5,r6,r7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "21e3bb0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['VF'], dtype='object')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3 = dataB[Name_sample2].str.split(':', expand=True)\n",
    "\n",
    "remove_empty_columns = len(df3.columns)-len(dataB.FORMAT.iloc[0].split(':'))\n",
    "df3 = df3.iloc[: , :-remove_empty_columns]\n",
    "\n",
    "df3.columns = dataB.FORMAT.iloc[0].split(':')\n",
    "dataB = pd.concat([dataB, df3], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "10d46f65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dataB.FORMAT.iloc[0].split(':'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a3f2208e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0/0</td>\n",
       "      <td>4,0</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>0,12,99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0/0</td>\n",
       "      <td>84,0</td>\n",
       "      <td>84</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0/1</td>\n",
       "      <td>84,79</td>\n",
       "      <td>163</td>\n",
       "      <td>99</td>\n",
       "      <td>1872,0,1877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0/0</td>\n",
       "      <td>201,0</td>\n",
       "      <td>201</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0/0</td>\n",
       "      <td>205,0</td>\n",
       "      <td>205</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1213</th>\n",
       "      <td>0/0</td>\n",
       "      <td>85,0</td>\n",
       "      <td>85</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1214</th>\n",
       "      <td>0/0</td>\n",
       "      <td>56,0</td>\n",
       "      <td>56</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>0/0</td>\n",
       "      <td>53,0</td>\n",
       "      <td>53</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216</th>\n",
       "      <td>0/0</td>\n",
       "      <td>89,0</td>\n",
       "      <td>89</td>\n",
       "      <td>99</td>\n",
       "      <td>0,120,1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>0/0</td>\n",
       "      <td>0,0</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>0,0,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1218 rows Ã 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0      1    2   3            4\n",
       "0     0/0    4,0    4  12      0,12,99\n",
       "1     0/0   84,0   84  99   0,120,1800\n",
       "2     0/1  84,79  163  99  1872,0,1877\n",
       "3     0/0  201,0  201  99   0,120,1800\n",
       "4     0/0  205,0  205  99   0,120,1800\n",
       "...   ...    ...  ...  ..          ...\n",
       "1213  0/0   85,0   85  99   0,120,1800\n",
       "1214  0/0   56,0   56  99   0,120,1800\n",
       "1215  0/0   53,0   53  99   0,120,1800\n",
       "1216  0/0   89,0   89  99   0,120,1800\n",
       "1217  0/0    0,0    0   .        0,0,0\n",
       "\n",
       "[1218 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_empty_columns = len(df3.columns)-len(dataB.FORMAT.iloc[0].split(':'))\n",
    "df3.iloc[: , :-remove_empty_columns]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faaf26b0",
   "metadata": {},
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e05958ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# NOOOOOOOO BORRARRRR ######################################\n",
    "\n",
    "\n",
    "### 6. Generate a report in a dashboard\n",
    "# Bootstrap themes by Ann: https://hellodash.pythonanywhere.com/theme_explorer\n",
    "app = dash.Dash(__name__, external_stylesheets=[dbc.themes.SIMPLEX])\n",
    "\n",
    "app.layout = dbc.Container([\n",
    "    dbc.Row([\n",
    "        dbc.Col([\n",
    "            dbc.Card([\n",
    "                dbc.CardImg(src='./assets/wrgllogohighres.jpeg')\n",
    "                ],color=\"light\", outline=True),\n",
    "                ],width=3),\n",
    "                dbc.Col([\n",
    "                    dbc.Card([\n",
    "                        dbc.CardImg(src='./assets/salisbury-nhs-foundation-trustLeft.png')\n",
    "                             ],color=\"light\", outline=True)\n",
    "                        ],width=2)\n",
    "            ],justify=\"between\"),\n",
    "        dbc.Row(\n",
    "            [dbc.Card([ \n",
    "                dbc.CardBody([\n",
    "                    html.Div([\n",
    "                    html.H1(\"Report: Do the samples belong to the same patient?\"),\n",
    "                    html.H3(\"This application takes common variants and compares the GT fields. A relative large number of common variants which GT fields match is indicative that both samples have a close relationship.\"),\n",
    "                ], style={'textAlign': 'center'}) \n",
    "                     ], style={'textLeft':'center', })      \n",
    "                      ], style={'height':'10vh'},color=\"light\", outline=True),\n",
    "        ]),\n",
    "        dbc.Row([\n",
    "            dbc.Card([ \n",
    "                dbc.CardBody([\n",
    "                    html.Div([\n",
    "                    html.H1(\"\"),\n",
    "                    html.H3(\"\"),\n",
    "                ], style={'textAlign': 'center'}) \n",
    "                     ], style={'textLeft':'center', })      \n",
    "                      ], style={'height':'10vh'},color=\"light\", outline=True),\n",
    "        ]),\n",
    "        dbc.Row([  \n",
    "            dbc.Col([\n",
    "            dash_table.DataTable(\n",
    "        id='datatable-interactivity',\n",
    "        columns=[\n",
    "            {\"name\": i, \"id\": i, \"deletable\": True, \"selectable\": True, \"hideable\": True}\n",
    "            if i == \"iso_alpha3\" or i == \"year\" or i == \"id\"\n",
    "            else {\"name\": i, \"id\": i, \"deletable\": True, \"selectable\": True}\n",
    "            for i in final_df.columns\n",
    "        ],\n",
    "        data=final_df.to_dict('records'),  # the contents of the table html.H3(os.path.basename(Name_file1)),\n",
    "        filter_action=\"native\",     # allow filtering of data by user ('native') or not ('none')\n",
    "        sort_action=\"native\",       # enables data to be sorted per-column by user or not ('none')\n",
    "        sort_mode=\"single\",         # sort across 'multi' or 'single' columns\n",
    "        row_deletable=True,         # choose if user can delete a row (True) or not (False)\n",
    "        selected_rows=[],           # indices of rows that user selects\n",
    "        page_action=\"native\",       # all data is passed to the table up-front or not ('none')\n",
    "        page_current=0,             # page number that user is on\n",
    "        style_cell={                # ensure adequate header width when text is shorter than cell's text\n",
    "            'minWidth': 200, 'maxWidth': 95, 'width': 95\n",
    "        },\n",
    "        style_data_conditional=[\n",
    "        {\n",
    "            'if': {'row_index': 'odd'},\n",
    "            'backgroundColor': 'rgb(248, 248, 248)'\n",
    "        }\n",
    "    ],\n",
    "        style_header={\n",
    "        'backgroundColor': 'rgb(230, 230, 230)',\n",
    "        'fontWeight': 'bold'\n",
    "    })]),\n",
    "    dbc.Col([\n",
    "        dbc.Card([ \n",
    "            dbc.CardBody([\n",
    "                html.Div([\n",
    "                    html.H4(\"Number of common variants:\",style={'color': 'black', 'fontSize': 15}),\n",
    "                    html.H4(len(final_df)),\n",
    "                    html.H4(\"Number of matches:\", style={'color': 'black', 'fontSize': 15}),\n",
    "                    html.H4(final_df.Matches.value_counts()[1]),\n",
    "                    ], \n",
    "                    style={'textAlign': 'left'})\n",
    "                    ])\n",
    "                    ],\n",
    "                    style={'height':'8vh'},color=\"light\", outline=True),\n",
    "                    ])\n",
    "                    ])\n",
    "                    ])\n",
    "\n",
    "\n",
    "\n",
    "  \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efec0d57",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
